### ISSUE TITLE
Guidance: basename params

### ISSUE THREAD

**Issue by BerkeleyTrue** [faceyspacey/redux-first-router#121]
Hey peeps! Fantastic project. I'm working on refactoring freeCodeCamp to use RFR in place of Redux Router. So far the transition has been very nice!

One issue I'm running into now is our language param.

We structure all our urls like so:

`/:lang/...`

This `lang` param can be defined but defaults to 'en'. We use a wrapper around create history to handle this logic for us [source](https://github.com/freecodecamp/freecodecamp/blob/staging/client/utils/use-lang-routes.js#L26).

This has worked great for us with RR.

The best part of this is that the rest of the app did not have to worry about this leading param. As the history module would ensure that it was always used. 

So when using RR Link's you can just ignore the lang param and RR would figure it out. 

i.e.
```js
<Link to='/challenges/html5/span' />
```
Would get turned into '/en/challenges/html5/span' or what ever the language selected was.

The issue now has to do with RFR's Links with action object `to`s. We lose the benefit of `useLangRoutes` here because RFR's actionToPath will crash and cause RFR to spit out a ton of warning into the counsel due to the lang param being undefined in the payload of the object.

I'm aware that the `to` param can take a string similar to RR but I'd love to also take advantage of colocating the logic that creates those links using the Object argument `to`.

Help me @faceyspacey, you're my only hope!

p.s. I'm working on this branch [here](https://github.com/berkeleytrue/freecodecamp/tree/feat/modern-challenge)

Here is what the warning looks like with the route action and routesMap expanded.

<img width="1273" alt="learn_to_code___freecodecamp" src="https://user-images.githubusercontent.com/6775919/30779912-20ac1748-a0b3-11e7-91a1-0cd31df15db6.png">

Also, this message prints out over 2000's times which really slows things down when trying to debug things.

**Comment by faceyspacey** [comment]
@BerkeleyTrue great to hear RFR is being used with freeCodeCamp!

I feel like if we can use the `basename` option to `createBrowser/MemoryHistory` it would solve a lot of things. I'm going to lay out how it's used and you let me know if that works for you:

So uust recently we added a `basename` option like `createMemory/BrowserHistory` have. Well in fact, `createMemoryHistory` doesn't have, but we have forked `history` and now we use it internally. Here's the fork: https://github.com/faceyspacey/rudy-history

So RFR will be renamed `rudy` and if you use the `@rudy` tag on NPM, you can get access to these features. If you check the top of this repo's readmes you will see updates about all the updates on the `@rudy` tag/branch. There are links to issue comments for how to implement things. 

But basically, the idea is you don't provide the `basename` option to `createBrowser/MemoryHistory`, but rather to the options of `connectRoutes`. Same goes for the `initialEntries` option if you're using `createMemoryHistory`. Here's how you implement:

```js
connectRoutes(routesMap, { 
   basename: '/awesome',
   initialEntries: [request.path], // only used on the server or in React Native
})
```
> `connectRoutes` internally detects if you're in a browser or an environment like Node or React Native that requires the memory history.

In addition, can provide your own `createHistory` factory like this:

import createBrowserHistory from 'rudy-history/createBrowserHistory'

```js
connectRoutes(routesMap, { 
   basename: '/awesome',
   createHistory: createBrowserHistory
})
```

I'm a little apprehensive that it will work for your setup or I assume you would have already been using it, but maybe there are other reasons with RR, and it will work nicely here. Here's how I imagine it for your setup:


```js
const location = typeof window !== 'undefined ? window.location.pathname : req.path
const basename = `/${getLangFromPath(location)}`

connectRoutes(routesMap, {  basename })
```

So that would generate URLs correctly with the basename. It doesn't answer the question why you didn't do this with RR + `history` to begin with. Does it have to do with a feature to switch languages or something like that? If so, I could add an RFR method to change the basename at any time. You would do this:

```js
import { changeBaseName } from 'redux-first-router'

class TopLevelAppComponent extends React.Component {
     onSelectLanguage = lang => {
          changeBaseName(lang) // will also automatically change URL to have basename
          this.forceUpdate() // all link components re-render correctly
     }
     render() {
         return (
            <div>
                <LanguageSelector onSelectLanguage={this.onSelectLanguage}
                <App />
            </div>
        )
     }
}
```

I mean wouldn't be nice to rid of the monkey-patching/customization of `createHistory` here:

https://github.com/freecodecamp/freecodecamp/blob/staging/client/utils/use-lang-routes.js#L26

I just need to understand your use-case better and I can easily tailor RFR to your needs so from your apps perspective different languages are handled completely idiomatically in terms of the router you use :)

Why do you want to manually add the basename to the paths you push here:

```js
const push = location => history.push(addLangToLocation(location, lang));
```
?

Why can't this be set it and forget it? Why do you have to stay ready to change the language here:

```js
 const unsubscribeFromHistory = history.listen(nextLocation => {
      lang = getLangFromLocation(nextLocation);
    });
```
?

Also, do you want to receive the `lang` in the payload? Doing it this way would essentially remove access to the language in your actions. You could add it as a 3rd argument to async functions using redux-thunk's extra argument feature. And you could add it to RFR's thunks similarly:


```js
const location = typeof window !== 'undefined ? window.location.pathname : req.path
const basename = `/${getLangFromPath(location)}`

const routesMap = {
     FOO: {
          // note: this is a pathless route; not all routes have to have paths
          // this allows for uniform thunks throughout your app
          thunk: (dispatch, getState, { extra: { basename } }) => ...
     }
}
connectRoutes(routesMap, {  basename, extra: { basename } })
```

So your reducers wouldn't get access to the basename though. However, it may in fact be a good idea to put the extra arg in actions passed to reducers so you could get it there. I could easily add that capability.

Either way, I can easily make it so your reducers and async actions (aka thunks) can have access to the basename so you don't have to import it globally, which is no good for SSR which requires the rendering for all requests to be isolated from each other (i.e. so no variables/state "leak" between requests).

That's all I got. In short, it would be nice to figure out how to make your setup use the `basename`. If the reason you didn't use it in the past was simply because `createMemoryHistory` didn't support a `basename` option--well, that's not a problem anymore with the `rudy-history` package. In an ideal world, your application in all but one place should be unaware of a basename. And perhaps for you that's 2 places since you likely will want asynchronously fetch language-specific content. The best option for that is probably to create your `api` instance so it sends all requests with the basename/lang in a header. So like this:



```js
import reduxThunk from 'redux-thunk'
import { connectRoutes } from 'redux-first-router'
import createApiArgument from './createApiArgument'

const location = typeof window !== 'undefined ? window.location.pathname : req.path
const lang = getLangFromPath(location)
const basename = `/${lang}`

const routesMap = {
     GET_LANG_SPECIFIC_CONTENT: {
         path: '/categories/:category',
          thunk: async (dispatch, getState, { extra: api }) => {
                const { category } = getState().location.payload
                const content = await api.fetchContent(category)
                dispatch({ type: 'NEW_CONTENT', payload: { content } })
         }
     }
}

const api = createApiArgument(lang)
const { middleware, enhancer, etc } = connectRoutes(routesMap, {  basename, extra: api })
const thunk = reduxThunk.withExtraArgument(api)
const middlewares = applyMiddleware(middleware, thunk)

// ...usual story
```

So from there `api.fetchContent` or whatever methods your `api` instance has would add the language to headers (or perhaps query params or post data, etc). And you would never need to individually apply it.

Let me know how I can best help you apply the basename/lang in as few places as possible so that RFR usage can remain idiomatic and unaware of the basename everywhere else. I.e. let me know how we can best apply a "single source of truth" strategy for you :)

**Comment by faceyspacey** [comment]
Also, thinking about it more, it seems the way you have it setup is so that for SEO the URLs embedded in the page never are prefixed with, say, `/en`. Rather they are always the same, but what gets pushed on the address bar (and what the rest of your app reads) is paths with the language prefixed. 

Is this an intentional SEO strategy so you don't have multiple links on the web? I couldn't actually seem to find any URLs in google prefixed with languages. So do they only appear client side, while the server interprets them if need be, but redirects or something to the URL without the language, but with the language-specific content rendered properly?

..Just trying to find the reason you might have wanted it that way. With the setup proposed, the URLs generated by the link component would all have the basename. Like, in the `href` attribute seen by google if server-rendered. Not just in what's pushed on the address bar.

**Comment by BerkeleyTrue** [comment]
> I feel like if we can use the basename option to createBrowser/MemoryHistory

I took a look at this and it doesn't seem like basename can be a param that RFR would interpret. I could be wrong though.

> So RFR will be renamed rudy

So the whole project is being renamed or is Rudy a codename?

> Why do you have to stay ready to change the language here

This is because the language of the page can be changed on the fly by the user.

> do you want to receive the lang in the payload?

I don't think we have a use for this. As long as it is available in `state.location` somewhere.

> this is a pathless route; not all routes have to have paths. this allows for uniform thunks throughout your app

This is pretty interesting. I'd like to learn more. Although I don't think this is the route we want to go for this issue. It should be pretty seemless and this seems to me to make the routes have to know about arbitrary logic unrelated to the specific routes.

> So your reducers wouldn't get access to the basename though

This is definitely what we don't want. We want RFR to manage this part of the state. It would be annoying to have url state in two different locations.

> your application in all but one place should be unaware of a basename

I think this is true of a basename but I'm not sure this is exactly like a basename. Basenames are meant to prefix your site to a uri subpath. This part of the uri is part of the site. This state should live right along side the rest of the params. It shouldn't be treated differently.

> Also, thinking about it more, it seems the way you have it setup is so that for SEO the URLs embedded in the page never are prefixed with, say, /en. Rather they are always the same, but what gets pushed on the address bar (and what the rest of your app reads) is paths with the language prefixed.

Yes this is correct. 

> Is this an intentional SEO strategy

Yes this is correct.

>  I couldn't actually seem to find any URLs in google prefixed with languages.

That's because this work is being done on our beta platform not yet released. You can find it [here](https://beta.freecodecamp.org).

Here is what the urls look like:
https://beta.freecodecamp.org/en/challenges/basic-javascript/storing-values-with-the-assignment-operator
https://beta.freecodecamp.org/es/challenges/basic-javascript/storing-values-with-the-assignment-operator

(Unfortunately it looks like our translations are not displaying ATM).

You've raised an interesting point. We want the href's on the page to ignore the lang param and just leave the rest of the URI. I might need to make our routesMap ignore the lang param in the path property. Will RFR still resolve URL's as long as the history module understands the first param?

**Comment by faceyspacey** [comment]
Ok we'll find how to do what ur trying to do. I'll respond more in depth later. But basically the standard thing that would be seemless is to have a basename or :lang param, while the embedded URLs actually show it. You're trying to hide the basename from embedded URLs, while pushing something different on the address bar:

That's definitely a unique strategy. And I think from an seo perspective it's great. But maybe it isn't--because since google will never see (or at least less often see) your non-English pages u will rank lower in foreign countries. Google probably wants to see that foreign language text more often than not.

But let's just say it's the best seo strategy, why even bother storing the language in the URL. Detect the location and store it in a cookie. All google ever sees is a shit ton of links to ur main English pages. That's essentially what ur trying to do.

Basically before proceeding the value of embedding links that are stripped of the basename must be fully scrutinized. It may be best to play less tricks on google while having easier to find language-specific URLs that are the same in all places. In which case either the basename or :lang strategy work fine. The basename one being preferable as u never have to worry about it after its setup. If u need to change the language midway a window.location = '/fr' for example would do the trick. And if u wanted to do it without a refresh, like I said I'd be more than happy to add a function called changeBaseName. I.e. so basename isn't just reserved for the initial evaluation of your store configuration code.

I'll answer specific questions as soon as I can.

**Comment by faceyspacey** [comment]
Subdomains or different extensions, e.g. .de instead of .com, might be another route worth looking into. That way the URLs are as clean as possible. And also relevant. For example German users likely would gravitate to freecodecamp.de far more anyway.

**Comment by faceyspacey** [comment]
Action/path transforms could work.

However if you are open to changing the embedded URLs, then that means everything works idiomatically with basenames. Then you can easily put the language in reducer state right when you configure the store. Then thunks, epics, etc can simply grab the language from state.

Then you just need a way to change the basename midway. A refresh would do it. At which point u don't need anything custom.

Either way, just so we are clear: is the goal now simply to get the Lang param in action.payload?

I'm trying to keep the API surface low--what I think might be great tho is if we put the basename in the action. Ie at action.meta.location.basename. We could also maintain it in the location reducer at: location.basename.

I think maintaining the basename is worthwhile information if available. This would make for a nice strategy to recommend to users that have multi-language sites. Ultimately I rather look at your use case as an opportunity to come up with the best way for multi-language sites to operate, instead of come up with good enough work arounds. RFR should have first class support for what ur trying to do, and ultimately and hopefully what ur trying to do should be a standard way to do such things. But it sounds like we are narrowing in on your use case being normal after all :)

**Comment by BerkeleyTrue** [comment]
> RFR should have first class support for what ur trying to do

Giving the user the ability to tap into these to API edges (action<>path logic) will give the user incredible power over the functionality of RFR. This will let me accomplish my goal as well as maybe open up RFR to other possibilities we haven't thought of without making it specifically about one use case.

**Comment by BerkeleyTrue** [comment]
Alright, I've got a version of Link working that pulls the current lang from url. I haven't tested changing langs yet but I'm sure it should work fine. 

I'm still interested in seeing the basename option working but I need to keep moving. I've looked through the rudy-history module and it looks like it's not set up to use a basename as a param. I'm wondering if this is something that could change.

Here is the new Link component: https://github.com/berkeleytrue/freecodecamp/blob/feat/modern-challenge/common/app/Router/Link.jsx#L35

**Comment by faceyspacey** [comment]
Brother, u set it in the options to connectRoutes. But rudy-history does take it too. Notes are at the top of this repo's readme for such new features.

...I'm not understanding u brother. I proposed an option to put the basename in actions and location state. I'm also still not sure if u have accepted the proposal to do ur links more regularly. I'm not sure why u would want to use a version of the link package or fork or monkeypatched wrapper when I can build u exactly what u want the correct way. But I don't know what u want because you aren't responding on recent key points :)

...I see ur putting the Lang in the payload. I was saying I can put the basename in actions and state for u and that's the same thing. And it would be the future proof way other developers would use that is promoted as the correct way.

**Comment by BerkeleyTrue** [comment]
I'm pretty happy with this solution and I don't see using `basename` as a huge improvement. 

I'm going to close this issue with the solution being to use my own version of Link. 

I appreciate the help and the time you put into this issue and this project. 

Happy Coding

**Comment by faceyspacey** [comment]
Ok let me know if u run into problems.
